File: src/creature-generation.js
import { createEmptyOriginalCreature, validateOriginalCreature } from "./creature-builder.js";

const clone = (value) => typeof structuredClone === "function" ? structuredClone(value) : JSON.parse(JSON.stringify(value));
const value = (object, camel, snake = camel) => object?.[camel] ?? object?.[snake];
const statistic = (number) => Number.isFinite(Number(number)) ? { band: "moderate", value: Number(number) } : null;

export class CreatureGenerationError extends Error {
  constructor(code, message, details = undefined) {
    super(message);
    this.name = "CreatureGenerationError";
    this.code = code;
    if (details) this.details = clone(details);
  }
}

/** Read-only semantic validation for Original and Forked Creature DTOs. */
export function validateCustomCreature(creature) {
  return validateOriginalCreature(clone(creature));
}

/**
 * Create an encounter-local Forked Creature from a complete Catalog Entry.
 * Mechanics and existing spellcasting blocks are copied; spell lists are
 * never generated at this boundary.
 */
export function forkExistingCreature(entry, { id = null, createdAt = "", origin = "webmcp" } = {}) {
  if (!entry || entry.kind !== "creature") throw new CreatureGenerationError("unsupported_catalog_entry", "Only Creature Catalog Entries can be forked.");
  if (entry.completeness !== "complete" || entry.support !== "supported") throw new CreatureGenerationError("catalog_entry_partial", "Only complete, supported Catalog Entries can be forked.");
  const detail = entry.detail ?? {};
  const role = entry.roles?.[0] ?? "brute";
  const blocks = [...(value(detail, "spellcastingBlocks", "spellcasting_blocks") ?? [])];
  const strikes = (detail.strikes ?? []).map((strike, index) => ({
    id: `strike_${index + 1}`,
    name: strike.name ?? "",
    actionCost: Number.isInteger(Number(strike.action_cost ?? strike.actionCost)) ? Number(strike.action_cost ?? strike.actionCost) : 1,
    traits: [...(strike.traits ?? [])],
    attack: statistic(strike.attack),
    damage: strike.damage ? [{ expression: String(strike.damage), type: "" }] : [],
    effect: ""
  }));
  const abilities = (detail.abilities ?? []).map((ability, index) => {
    const rawCost = value(ability, "actionCost", "action_cost");
    const kind = ["reaction", "free_action", "passive"].includes(String(rawCost)) ? String(rawCost) : "action";
    return {
      id: `ability_${index + 1}`,
      name: ability.name ?? "",
      kind,
      actionCost: [1, 2, 3].includes(Number(rawCost)) ? Number(rawCost) : null,
      traits: [...(ability.traits ?? [])],
      trigger: null,
      requirements: null,
      target: null,
      range: null,
      area: null,
      resolution: null,
      damage: [],
      conditions: [],
      duration: null,
      frequency: null,
      effectText: ability.text ?? ""
    };
  });
  const creature = createEmptyOriginalCreature();
  creature.id = id ?? `cre_forked_${String(entry.content_id).replace(/[^a-z0-9]+/gi, "_").replace(/^_|_$/g, "").toLowerCase()}`;
  creature.identity = {
    name: entry.name ?? "",
    level: Number(entry.level),
    rarity: entry.rarity ?? "common",
    size: ({ med: "medium", sm: "small", lg: "large", huge: "huge" })[detail.size] ?? detail.size ?? "medium",
    traits: [...(entry.traits ?? [])],
    concept: detail.concept ?? entry.summary ?? "",
    roadmap: role,
    encounterRole: role
  };
  creature.perception = statistic(detail.perception);
  creature.senses = [...(detail.senses ?? [])];
  creature.languages = [...(detail.languages ?? [])];
  creature.skills = { ...(detail.skills ?? {}) };
  creature.defenses = {
    ac: statistic(detail.defenses?.ac),
    fortitude: statistic(detail.defenses?.fortitude),
    reflex: statistic(detail.defenses?.reflex),
    will: statistic(detail.defenses?.will),
    hp: statistic(detail.defenses?.hp),
    immunities: [], weaknesses: [], resistances: []
  };
  creature.speeds = { ...(detail.speeds ?? {}) };
  creature.strikes = strikes;
  creature.abilities = abilities;
  creature.spellcastingStatus = blocks.length || entry.spellcasting ? "preserved_existing" : "none";
  creature.spellcastingBlocks = blocks;
  creature.tactics = detail.tactics ?? "";
  creature.morale = detail.morale ?? "";
  creature.provenance = { origin: "forked", basedOnContentID: entry.content_id, createdAt, mutationOrigin: origin };
  return creature;
}

/** Validate, then return a detached commit snapshot for the host mutation. */
export function commitCustomCreature(creature, { origin = "webmcp" } = {}) {
  const snapshot = clone(creature);
  const readiness = validateCustomCreature(snapshot);
  if (readiness.structuralErrors.length > 0) {
    throw new CreatureGenerationError("creature_structural_errors", "The Creature has structural errors.", { fields: readiness.structuralErrors.map((item) => item.field) });
  }
  snapshot.provenance = { ...(snapshot.provenance ?? {}), origin: snapshot.provenance?.origin || "original", mutationOrigin: origin };
  return snapshot;
}

File: native/Sources/SidekickDMCore/CreatureGeneration.swift
import Foundation

/// Errors at the semantic creature-generation boundary. The command host is
/// responsible for translating these into the protocol's native error envelope.
public enum CreatureGenerationError: Error, Equatable, Sendable {
    case unsupportedCatalogEntry
    case catalogEntryPartial
    case structural([CreatureValidationIssue])

    public var code: String {
        switch self {
        case .unsupportedCatalogEntry: return "unsupported_catalog_entry"
        case .catalogEntryPartial: return "catalog_entry_partial"
        case .structural: return "creature_structural_errors"
        }
    }

    public var message: String {
        switch self {
        case .unsupportedCatalogEntry: return "Only Creature Catalog Entries can be forked."
        case .catalogEntryPartial: return "Only complete, supported Catalog Entries can be forked."
        case .structural: return "The Creature has structural errors."
        }
    }
}

/// Pure transformations used by semantic WebMCP creature tools. Keeping
/// validation and catalog conversion here means a read-only validation call
/// cannot accidentally touch the encounter store; the host owns the eventual
/// revision-checked atomic mutation.
public enum CreatureGeneration {
    public static func validate(_ creature: OriginalCreature) -> CreatureValidationResult {
        CreatureBuilder.validate(creature)
    }

    public static func commit(_ creature: OriginalCreature, origin: String = "webmcp") throws -> OriginalCreature {
        do {
            return try CreatureBuilder.create(creature, origin: origin)
        } catch let error as CreatureBuilderError {
            if case .structural(let issues) = error { throw CreatureGenerationError.structural(issues) }
            throw error
        }
    }

    /// Creates an encounter-local Forked Creature from a complete Catalog
    /// Creature. All catalog mechanics, including existing spellcasting
    /// blocks, are copied as-is. Generation never synthesizes a spell list.
    public static func fork(
        _ catalogCreature: CatalogCreature,
        id: String? = nil,
        createdAt: String = "",
        origin: String = "webmcp"
    ) throws -> OriginalCreature {
        guard catalogCreature.summary.kind == .creature else { throw CreatureGenerationError.unsupportedCatalogEntry }
        guard catalogCreature.summary.completeness == .complete, catalogCreature.summary.support == .supported else {
            throw CreatureGenerationError.catalogEntryPartial
        }

        let role = catalogCreature.summary.roles.first ?? .brute
        let defenses = CreatureDefenses(
            ac: statistic(catalogCreature.defenses["ac"]),
            fortitude: statistic(catalogCreature.defenses["fortitude"]),
            reflex: statistic(catalogCreature.defenses["reflex"]),
            will: statistic(catalogCreature.defenses["will"]),
            hp: statistic(catalogCreature.defenses["hp"])
        )
        let strikes = catalogCreature.strikes.enumerated().map { index, strike in
            CreatureStrike(
                id: "strike_\(index + 1)",
                name: strike.name,
                actionCost: 1,
                traits: strike.traits,
                attack: statistic(strike.attack),
                damage: strike.damage.map { [CreatureDamage(expression: $0)] } ?? []
            )
        }
        let abilities = catalogCreature.abilities.enumerated().map { index, ability in
            CreatureAbility(
                id: "ability_\(index + 1)",
                name: ability.name,
                kind: abilityKind(ability.actionCost),
                actionCost: actionCost(ability.actionCost),
                traits: ability.traits,
                effectText: ability.text
            )
        }
        let blocks = catalogCreature.spellcastingBlocks
        let spellcastingStatus = blocks.isEmpty && catalogCreature.summary.spellcasting != true ? "none" : "preserved_existing"
        return OriginalCreature(
            id: id ?? "cre_forked_\(slug(catalogCreature.summary.contentID))",
            identity: CreatureIdentity(
                name: catalogCreature.summary.name,
                level: catalogCreature.summary.level,
                rarity: catalogCreature.summary.rarity,
                size: catalogCreature.size,
                traits: catalogCreature.summary.traits,
                concept: catalogCreature.concept,
                roadmap: roadmap(for: role),
                encounterRole: role
            ),
            perception: statistic(catalogCreature.perception),
            senses: catalogCreature.senses,
            languages: catalogCreature.languages,
            skills: catalogCreature.skills,
            defenses: defenses,
            speeds: catalogCreature.speeds,
            strikes: strikes,
            abilities: abilities,
            spellcastingStatus: spellcastingStatus,
            spellcastingBlocks: blocks,
            tactics: catalogCreature.tactics,
            morale: catalogCreature.morale,
            provenance: CreatureProvenance(origin: "forked", basedOnContentID: catalogCreature.summary.contentID, createdAt: createdAt, mutationOrigin: origin)
        )
    }

    private static func statistic(_ value: Int?) -> CreatureStatistic? {
        guard let value else { return nil }
        return CreatureStatistic(band: .moderate, value: value)
    }

    private static func roadmap(for role: EncounterRole) -> CreatureRoadmap? {
        CreatureRoadmap(rawValue: role.rawValue)
    }

    private static func actionCost(_ value: String?) -> Int? {
        switch value?.lowercased() {
        case "1", "one", "action": return 1
        case "2", "two": return 2
        case "3", "three": return 3
        default: return nil
        }
    }

    private static func abilityKind(_ value: String?) -> CreatureAbilityKind {
        switch value?.lowercased() {
        case "reaction": return .reaction
        case "free_action", "free action": return .freeAction
        case "passive": return .passive
        default: return .action
        }
    }

    private static func slug(_ value: String) -> String {
        let result = value.lowercased().map { character in
            character.isLetter || character.isNumber ? String(character) : "_"
        }.joined()
        return result.trimmingCharacters(in: CharacterSet(charactersIn: "_"))
    }
}
